File: src/FinanceTracker.Core/Models/Transaction.cs
namespace FinanceTracker.Core.Models;

public enum TransactionType { Income, Expense }

public class Transaction
{
    public Guid Id { get; set; }
    public string Description { get; set; } = string.Empty;
    public decimal Amount { get; set; }
    public TransactionType Type { get; set; }
    public DateTime Date { get; set; }
    public string Currency { get; set; } = "EUR";
    public string? Notes { get; set; }
    public string? ExternalAccountId { get; set; }
    public string? ExternalId { get; set; }

    // Open Banking
    public string? ExternalTransactionId { get; set; }

    public Guid AccountId { get; set; }
    public Account Account { get; set; } = null!;

    public Guid? CategoryId { get; set; }
    public Category? Category { get; set; }
}

File: finance-tracker-ui/src/app/services/transaction.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface Transaction {
  id: string;
  description: string;
  amount: number;
  type: 'Income' | 'Expense';
  date: string;
  currency: string;
  notes?: string;
  accountId: string;
  account?: { id: string; name: string };
  categoryId?: string;
  category?: { id: string; name: string; icon: string; color: string };
}

export interface CreateTransactionDto {
  description: string;
  amount: number;
  type: 'Income' | 'Expense';
  date: string;
  currency: string;
  notes?: string;
  accountId: string;
  categoryId?: string;
}

@Injectable({ providedIn: 'root' })
export class TransactionService {
  private http = inject(HttpClient);
  private apiUrl = 'http://localhost:5097/api/transactions';

  getAll(): Observable<Transaction[]> {
    return this.http.get<Transaction[]>(this.apiUrl);
  }

  create(dto: CreateTransactionDto): Observable<Transaction> {
    return this.http.post<Transaction>(this.apiUrl, dto);
  }

  delete(id: string): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}